src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentStepExecution.java

```diff
@@ -0,0 +1,128 @@
+package com.atlassian.bitbucket.jenkins.internal.deployments;
+
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.DeploymentState;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepository;
+import com.atlassian.bitbucket.jenkins.internal.status.BitbucketRevisionAction;
+import com.google.common.annotations.VisibleForTesting;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepExecutionImpl;
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+
+import java.util.logging.Logger;
+
+public class DeploymentStepExecution extends AbstractStepExecutionImpl {
+
+    private static final long serialVersionUID = 1L;
+
+    private final transient BitbucketDeploymentFactory deploymentFactory;
+    private final transient DeploymentPoster deploymentPoster;
+    private final BitbucketDeploymentEnvironment environment;
+
+    DeploymentStepExecution(BitbucketDeploymentEnvironment environment,
+                            StepContext context) {
+        this(environment, context, getStepDescriptor().getBitbucketDeploymentFactory(), getStepDescriptor().getDeploymentPoster());
+    }
+
+    @VisibleForTesting
+    DeploymentStepExecution(BitbucketDeploymentEnvironment environment,
```

Comment:
It looks like this has been done to allow mock injection in the test methods, but maybe you could just make the `getStepDescriptor` a public non-static method (for which there is precedent in Jenkins, e.g. `SCMSource.getDescriptor`) and your test classes could construct the deployment step as a spy? `BitbucketSCMSourceTest.createInstance` does something very similar. This is okay as it is but I try to avoid @VisibleForTesting as much as possible, especially if the unit tests can carry the messiness instead.